source/api-abi-versioning/abi-versioning.txt

```diff
@@ -0,0 +1,173 @@
+   do not support compatibility of the public API *across* different build
+   configurations. For example, a shared library generated with
+   ``BSONCXX_OUTPUT_NAME=bsoncxx`` is not expected to be compatible with a
+   program compiled against a stable ABI produced using a build configuration
+   with ``BSONCXX_OUTPUT_NAME=bsoncxx-custom-basename``. (This is particularly
+   important when using multi-config generators such as Visual Studio!)
+   
+.. _cpp-abi-root-namespace-redec:
+
+Root Namespace Redeclarations
+-----------------------------
+
+The root namespace provides "redeclarations" of entities declared in an ABI namespace. The ABI namespace may differ across entities being redeclared.
+
+For example, all of the following redeclarations may be present simultaneously:
+
+- ``bsoncxx::example::foo`` -- ``bsoncxx::v_noabi::example::foo``
+- ``bsoncxx::example::bar`` -- ``bsoncxx::v1::example::bar``
+- ``bsoncxx::example::baz`` -- ``bsoncxx::v2::example::baz``
+
+Root namespace redeclarations are designed to allow for the addition of binary incompatible symbols in ``vB`` without breaking binary compatibility of ``vA`` symbols.
+They facilitate the deprecation of stable ABI symbols while providing the opportunity for a clean transition from ``vA`` to ``vB`` without breaking binary compatibility.
+They allow user code to opt into "redeclaration upgrades" that reduces required changes to source code when transitioning from ``vA`` to ``vB``.
+
+These redeclaration upgrades are intended to support a clean transition away from deprecated, to-be-removed ABI symbols across releases.
+Accordingly, users are encouraged to use the root namespace declarations by default to opt into these upgrades.
+However, users may need to use explicit ABI namespace qualifiers when referencing C++ Driver entities in their own stable ABIs depending on their own stability policy.
+
+The following compatibility table describes when the API major version number or ABI version number must be bumped due to an incompatible change:
+
+.. list-table::
+   :header-rows: 0
+   :stub-columns: 1
+
+   * - Change Type
+     - Source Compatible
+     - Binary Compatible
+
+   * - Add a new ``vB`` symbol
+     - Yes
+     - Yes
+
+   * - Upgrade a redeclaration from ``vA`` to ``vB``
+     - Maybe (1)
+     - Yes
+
+   * - Remove a ``vA`` symbol (from the public API)
+     - No
+     - Maybe (2)
+
+   * - Remove a ``vA`` symbol (from the stable ABI)
+     - Maybe (2)
+     - No
+
+- (1): A ``vB`` entity could be 100% source compatible with the ``vA`` API despite requiring use of a new, incompatible stable ABI symbol.
+- (2): A stable ABI symbol could be removed from the public API (e.g. via documentation or removal from public headers) while still providing an exported symbol definition for backward compatibility. This is probably unlikely to happen, but if it does, it would be a rare case where the source and binary incompatible changes can be split into separate releases.
```

Comment:
These two list items are meant to be "footnotes". If there is a better way to implement this as a footnote, recommend applying it here.

Reply:
I played around with this, but we're pretty limited with formatting couldn't come up with anything better.

Reply:
If the documentation generator is using ReStructuredText, it should support [inline footnotes](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#footnotes):

```rst
.. list-table::
   :header-rows: 0
   :stub-columns: 1

   * - Change Type
     - Source Compatible
     - Binary Compatible

   * - Add a new ``vB`` symbol
     - Yes
     - Yes

   * - Upgrade a redeclaration from ``vA`` to ``vB``
     - Maybe [1]_
     - Yes

   * - Remove a ``vA`` symbol (from the public API)
     - No
     - Maybe [2]_

   * - Remove a ``vA`` symbol (from the stable ABI)
     - Maybe [2]_
     - No

.. [1] A ``vB`` entity could be 100% source compatible with the ``vA`` API despite requiring use of a new, incompatible stable ABI symbol.

.. [2] A stable ABI symbol could be removed from the public API (e.g. via documentation or removal from public headers) while still providing an exported symbol definition for backward compatibility. This is probably unlikely to happen, but if it does, it would be a rare case where the source and binary incompatible changes can be split into separate releases.
```